src/realm/sync/noinst/client_reset.cpp

```diff
@@ -426,110 +446,203 @@ void remove_pending_client_resets(Transaction& wt)
+    else if (mode == 1) {
+        return ClientResyncMode::Recover;
+    }
+    else {
+        throw ClientResetFailed(util::format("Unsupported client reset metadata mode: %1 for pending reset", mode));
+    }
+}
+
+int64_t from_resync_mode(ClientResyncMode mode)
+{
+    switch (mode) {
+        case ClientResyncMode::Recover:
+            [[fallthrough]];
+        case ClientResyncMode::RecoverOrDiscard:
+            return 1; // Recover
+        case ClientResyncMode::DiscardLocal:
+            return 0; // Discard
+        default:
+            throw ClientResetFailed(
+                util::format("Unsupported client reset metadata mode: %1 for pending reset", mode));
+    }
+}
+
 util::Optional<PendingReset> has_pending_reset(const Transaction& rt)
 {
     ConstTableRef table = rt.get_table(s_meta_reset_table_name);
     if (!table || table->size() == 0) {
         return util::none;
     }
-    ColKey timestamp_col = table->get_column_key(s_timestamp_col_name);
-    ColKey type_col = table->get_column_key(s_reset_type_col_name);
-    ColKey version_col = table->get_column_key(s_version_column_name);
-    REALM_ASSERT(timestamp_col);
-    REALM_ASSERT(type_col);
-    REALM_ASSERT(version_col);
     if (table->size() > 1) {
         // this may happen if a future version of this code changes the format and expectations around reset metadata.
         throw ClientResetFailed(
             util::format("Previous client resets detected (%1) but only one is expected.", table->size()));
     }
+    ColKey version_col = table->get_column_key(s_version_column_name);
+    ColKey timestamp_col = table->get_column_key(s_timestamp_col_name);
+    ColKey mode_col = table->get_column_key(s_reset_mode_col_name);
     Obj first = *table->begin();
     REALM_ASSERT(first);
     PendingReset pending;
-    int64_t version = first.get<int64_t>(version_col);
+    REALM_ASSERT(version_col);
+    pending.version = static_cast<int>(first.get<int64_t>(version_col));
+    // Version 1 columns
+    REALM_ASSERT(timestamp_col);
+    REALM_ASSERT(mode_col);
     pending.time = first.get<Timestamp>(timestamp_col);
-    if (version > metadata_version) {
-        throw ClientResetFailed(util::format("Unsupported client reset metadata version: %1 vs %2, from %3", version,
-                                             metadata_version, pending.time));
-    }
-    int64_t type = first.get<int64_t>(type_col);
-    if (type == 0) {
-        pending.type = ClientResyncMode::DiscardLocal;
+    pending.mode = to_resync_mode(first.get<int64_t>(mode_col));
+    if (pending.version == 0 || pending.version > s_metadata_version) {
+        throw ClientResetFailed(util::format("Unsupported client reset metadata version: %1 vs %2, from %3",
+                                             pending.version, s_metadata_version, pending.time));
     }
-    else if (type == 1) {
-        pending.type = ClientResyncMode::Recover;
+    // Version 2 columns
+    if (pending.version >= 2) {
```

Comment:
Do we want to use the [sync metadata schema](https://github.com/realm/realm-core/blob/e2be13f6889f7c0bcf50fd05d6d042d17a81b02e/src/realm/sync/noinst/sync_metadata_schema.hpp#L123) machinery here to create all these column/look up these colkeys for us?

Reply:
Migrated the pending reset function to use the sync metadata schema and made it a "store" object that is owned by `SyncSession`.